sockeye/inference.py

```diff
@@ -881,6 +932,21 @@ def _beam_search(self,
         # reset all padding distribution cells to np.inf
         self.pad_dist[:] = np.inf
 
+        # If using a top-k lexicon, select param rows for logit computation that correspond to the
+        # target vocab for this sentence.
+        models_output_layer_w = list()
+        models_output_layer_b = list()
+        pad_dist = self.pad_dist
+        vocab_slice_ids = None  # type: np.ndarray
+        if self.restrict_lexicon:
+            # TODO: See note in method about migrating to pure MXNet when operations are supported.
+            vocab_slice_ids = self.restrict_lexicon.get_trg_ids(source.astype("int32").asnumpy())
+            pad_dist = mx.nd.full((self.batch_size * self.beam_size, vocab_slice_ids.shape[0]),
+                                  val=np.inf, ctx=self.context)
+            for m in self.models:
+                models_output_layer_w.append(mx.nd.array(m.output_layer_w[vocab_slice_ids], ctx=self.context))
```

Comment:
The reason you are saving <model>.output_layer_{w,b} as numpy arrays is because of these two lines because you have a list of indices in vocab_slice_ids, correct?
Instead of re-creating the mx.nd.array here every single time you can also use MXNet's `take` operation for this, if m.output_layer_{w,b} is NDArray and `lexicon.get_trg_ids` would return NDArray.
Even in the new 'fluent' mode:
```
...apend(m.output_layer_w.take(vocab_slice_ids))
```
Copying the vocab_slice_ids to the context is probably cheaper than copying the weight and bias parameters back and forth.

Reply:
Good catch--that's the op I was looking for.  It would be more natural if NDArray supported slicing by arrays like NumPy, but I see why they would want to use `take` for consistency between nd and sym APIs.

Reply:
I agree, numpy-like indexing and item assignment would be very useful to have. I think its on their roadmap for MXNet 0.13(?).
This just caused me some trouble as well here: #187 :)